### PR TITLE
Refactor: Unify EthTx to FilecoinMessage methods

### PR DESCRIPTION
**Pull request by arajasek**
## Related Issues

#10013 

## Proposed Changes

Unifies the common message decoding logic, making `newEthTxFromFilecoinMessage` a much simpler method.

## Checklist

Before you mark the PR ready for review, please make sure that:

- [ ] Commits have a clear commit message.
- [ ] PR title is in the form of of `<PR type>: <area>: <change being made>`
  - example: ` fix: mempool: Introduce a cache for valid signatures`
  - `PR type`: fix, feat, build, chore, ci, docs, perf, refactor, revert, style, test
  - `area`, e.g. api, chain, state, market, mempool, multisig, networking, paych, proving, sealing, wallet, deps
- [ ] New features have usage guidelines and / or documentation updates in
  - [ ] [Lotus Documentation](https://lotus.filecoin.io)
  - [ ] [Discussion Tutorials](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [ ] Tests exist for new functionality or change in behavior
- [ ] CI is green

### CODE REVIEW

**Comment by arajasek** [comment]
Test failure is known, and will be fixed in #10005 by @jennijuju by updating bundle

**Comment by geoff-vball** [comment]
Hey, as-is this doesn’t work yet. (But is much much cleaner thank you thank you).

I have two tests written that you can copy into your PR and test this with (just comment out the one line that will show up as red). I ran them and they’re both broken.

 https://github.com/filecoin-project/lotus/blob/c5c19dbf1d7d1cd0a28307ee56a257e7cd6c7bc5/itests/eth_hash_lookup_test.go#L196

https://github.com/filecoin-project/lotus/blob/c5c19dbf1d7d1cd0a28307ee56a257e7cd6c7bc5/itests/eth_hash_lookup_test.go#L266

Note that BLS messages have always been broken, but secp was working before your refactor. If you want to try to wrangle these you can, otherwise I’ll rebase by PR onto yours and try to make it work, probably in the wee hours of the morning tomorrow.

File: node/impl/full/eth.go [hunk 1]
```diff
@@ -1455,77 +1455,23 @@ func lookupEthAddress(ctx context.Context, addr address.Address, sa StateAPI) (e
...
 	if err != nil {
-		return ethtypes.EthTx{}, err
+		return ethtypes.EthTx{}, xerrors.Errorf("failed to convert from signed message: %w", err)
 	}
 
-	toAddr := &toEthAddr
-	input := smsg.Message.Params
-	// Check to see if we need to decode as contract deployment.
-	// We don't need to resolve the to address, because there's only one form (an ID).
-	if smsg.Message.To == builtintypes.EthereumAddressManagerActorAddr {
```
Comment (Stebalien) [inline node/impl/full/eth.go:1472]: So, the downside of unifying this is that we'll no longer be able to interpret "native" messages as ethereum transactions. On the other hand... that's probably not desirable anyways.

File: node/impl/full/eth.go [hunk 2]
```diff
@@ -1455,77 +1455,23 @@ func lookupEthAddress(ctx context.Context, addr address.Address, sa StateAPI) (e
...
-	// Otherwise, try to decode as a cbor byte array.
-	// TODO: Actually check if this is an ethereum call. This code will work for demo purposes, but is not correct.
-	if toAddr != nil {
-		if decodedParams, err := cbg.ReadByteArray(bytes.NewReader(smsg.Message.Params), uint64(len(smsg.Message.Params))); err == nil {
-			input = decodedParams
-		}
-	}
-
-	r, s, v, err := ethtypes.RecoverSignature(smsg.Signature)
+	fromEthAddr, err := lookupEthAddress(ctx, smsg.Message.From, sa)
```
Comment (Stebalien) [inline node/impl/full/eth.go:1463]: Given my comment below, I don't think we need this anymore. We _should_ be able to just use the `From` field from the message as-is, as it _must_ be an f410 address for the transaction to be "valid".

File: node/impl/full/eth.go [hunk 3]
```diff
@@ -1455,77 +1455,23 @@ func lookupEthAddress(ctx context.Context, addr address.Address, sa StateAPI) (e
...
 	if err != nil {
-		// we don't want to return error if the message is not an Eth tx
-		r, s, v = ethtypes.EthBigIntZero, ethtypes.EthBigIntZero, ethtypes.EthBigIntZero
+		return ethtypes.EthTx{}, xerrors.Errorf("failed to lookup from ethaddress: %w", err)
 	}
 
-	hash, err := ethtypes.EthHashFromCid(smsg.Cid())
+	toEthAddr, err := lookupEthAddress(ctx, smsg.Message.To, sa)
```
Comment (Stebalien) [inline node/impl/full/eth.go:1468]: If this is a valid eth transaction, `to` should either be an ID address, or an f410 address. We shouldn't need to resolve this if we're going to be limiting support to _real_ eth transactions.

**Review by Stebalien — COMMENTED** [review]
In-line with @geoff-vball's comments, the code here used to work with native accounts and now it doesn't. However, to be honest, we might not want it to work with native accounts anyways? I'm not sure.

File: node/impl/full/eth.go (hunk 1 above)
Comment (raulk) [inline node/impl/full/eth.go:1472]: Doesn't this change lose the ability to reverse engineer EVM contract creations performed from EEOAs? What's the motivation here?

**Review by raulk — CHANGES_REQUESTED** [review]
Let's document all the difficulties in returning native messages, e.g.:

- Inability to deal with non-existent f1 and f3 recipient addresses. Not a problem for executed messages (who have acquired an ID address), but a problem for `pending` messages.
- Inability to extract BLS signatures. This is fine, just set these fields to the zero value.
- Cannot convey the method number as a dedicated field. Can instead prepend it to the input params, Solidity-style by convention (need to document that convention).
- No possibility to communicate a message version number. We haven't ever moved from 0, so don't expect that to be a problem. Any future message schemas would need to map to the Eth schema anyway.

We have two options here:
1. Solve for these difficulties, and retain the ability to return non-Eth messages.
2. Drop this functionality.

A grey area solution that returns inaccurate data or incomplete data is not safe.

Reasons we want to support it:
1. Doesn't require Eth clients to learn a new API (Filecoin API, which doesn't have js libraries as mature as Eth).
2. Doesn't cause fragmentation in the API (some APIs deal with one part of the chain, other APIs with other parts of the chain), which is consistent with our interoperability principle and avoidance of hyperspecialization.

**Comment by arajasek** [comment]
We're gonna block this until https://github.com/filecoin-project/lotus/pull/9965 lands, then land into https://github.com/filecoin-project/lotus/pull/10026 at which point it'll be good for review.

**Comment by arajasek** [comment]
Superseded by #10095

**Pull request closed without merging**